source/detail/parser/x86_64.cpp

```diff
@@ -31,6 +32,98 @@ namespace smeagle::x86_64 {
   };
 
+  // A RegisterAllocator can provide the next register location
+  class RegisterAllocator {
+  public:
+    RegisterAllocator();
+    std::string getRegisterString(RegisterClass regClass, st::Type *paramType);
+    std::string getRegistersString(std::pair<RegisterClass, RegisterClass> regClasses,
+                                   st::Type *paramType);
+
+  private:
+    std::stack<std::string> intRegisters{{"%r9", "%r8", "%rcx", "%rdx", "%rsi", "%rdi"}};
+    std::stack<std::string> sseRegisters;
+    std::string getNextIntRegister();
+    std::string getNextSseRegister();
+    int nextMultipleEight(int number);
+    int framebase = 8;
+    void updateFramebaseFromType(st::Type *paramType);
+  };
+
+  // Constructor
+  RegisterAllocator::RegisterAllocator() {
+    // Populate the sse register stack
+    for (int i = 1; i < 8; ++i) {
+      sseRegisters.push("xmm" + i);
+    }
+  }
+
+  // Get the next greater multiple of 8
+  int RegisterAllocator::nextMultipleEight(int number) { return ((number + 7) & (-8)); }
+
+  // Get a framebase for a variable based on stack location and type
+  void RegisterAllocator::updateFramebaseFromType(st::Type *paramType) {
+    framebase += nextMultipleEight(paramType->getSize());
+  }
+
+  // Get the next available integer register
+  std::string RegisterAllocator::getNextIntRegister() {
+    // If we are empty, return stack
+    if (intRegisters.empty()) {
+      return "stack";
+    }
+    std::string regString = intRegisters.top();
+    intRegisters.pop();
+    return regString;
+  }
+
+  // Get the next available integer register
+  std::string RegisterAllocator::getNextSseRegister() {
+    // If we are empty, return stack
+    if (intRegisters.empty()) {
+      return "stack";
+    }
+    std::string regString = sseRegisters.top();
+    sseRegisters.pop();
+    return regString;
+  }
+
+  // Given two registers, return one combined string
+  std::string RegisterAllocator::getRegistersString(
+      std::pair<RegisterClass, RegisterClass> regClasses, st::Type *paramType) {
+    std::string locA = this->getRegisterString(regClasses.first, paramType);
+    std::string locB = this->getRegisterString(regClasses.second, paramType);
+
+    if (locB == "") {
+      return locA;
+    }
+    return locA + "|" + locB;
+  }
+
+  // Get a string location from a register class
+  std::string RegisterAllocator::getRegisterString(RegisterClass regClass, st::Type *paramType) {
+    std::string regString;
+
+    // If the class is memory, pass the argument on the stack
+    if (regClass == RegisterClass::NO_CLASS) return "";
+    if (regClass == RegisterClass::SSE)
+      regString = this->getNextSseRegister();
+    else if (regClass == RegisterClass::INTEGER)
+      regString = this->getNextIntRegister();
+    else if (regClass == RegisterClass::MEMORY)
+      regString = "stack";
+
+    // If we've run out of registers or have RegisterClass::MEMORY
+    if (regString == "stack") {
```

Comment:
It's your codebase, so you can leave this as-is, but using strings to communicate state in C++ is strongly discouraged. The real power of C++ lies in its type system. Exploiting types to your advantage makes for easier code to read, write, and maintain (I've heard it also makes you super cool :D ). Here, I might try a `std::optional<string>` to signal if register allocation succeeded or not.

Reply:
I gave a shot with the std::optional<std::string> !

Reply:
Oh, yeah, that's much better.